File: Exercise_2.py
# ## Problem2
# Search in a Rotated Sorted Array (https://leetcode.com/problems/search-in-rotated-sorted-array/)
# // Time Complexity : O(logm + logn) = O(logmn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach in three sentences only
class Solution:
    def search(self, reader, target):
        # defining 2 ptrs low
        low = 0
        # and high
        high = 1
        # till target comes in range
        while (reader.get(high) < target):
            # low to high
            low = high
            # increasing range by 2 times
            high = 2 * high
        # binary search
        while (low <= high):
            # calculating mid
            mid = (low + high)//2
            # if mid has target
            if reader.get(mid) == target:
                # target in the array
                return mid
            # left part of the array
            elif target >= reader.get(low) and target < reader.get(mid):
                # updating high - shortening search by 1/2
                high = mid - 1
            else:
                # updating low - shortening search by 1/2
                low = mid + 1
        return -1

File: Exercise_2.cpp
// Time Complexity                                  : O(N)
// Space Complexity                                 : O(Log N)
// Did this code successfully run on Leetcode       : Yes.
// Any problem you faced while coding this          : Faced issues with the edge cases. 
//                                                    1. When the array has not been rotated, I did not consider this case initially.               

#include<bits/stdc++.h>
using namespace std;

class Solution {
private: 
    int n;
    int getP(vector<int>& nums) {
        if(nums[0] < nums[n-1]) return 0;
        
        int l=0,r=n-1,md;
        while(l<=r) {
            md = l+(r-l)/2;
            
            if(nums[md]>nums[md+1]) 
                return md+1;
            else {
                if(nums[md]>=nums[0]) 
                    l = md+1;
                else 
                    r=md-1;
            }
        }
        return 0;
    }
    
    int srch(vector<int>& nums, int target, int l, int r) {
        int md;
        while(l<=r){
            md=l+(r-l)/2;
            if(nums[md]==target) return md;
            else if(target > nums[md]) l = md+1;
            else r = md-1;
        }
        return -1;
    }
    
public:
    int search(vector<int>& nums, int target) {
        n = nums.size();
        if(n==1){
            if(nums[0]==target) return 0;
            else return -1;
        }
        
        int p=getP(nums);
        
        if(nums[p] == target)
            return p;
        if(p == 0)
            return srch(nums,target,0,n-1);
        
        if(target >= nums[0])
            return srch(nums,target,0,p-1);
        else
            return srch(nums,target,p,n-1);
        
        return -1;
        
    }
};